core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java

```diff
@@ -0,0 +1,603 @@
+        if (fieldDataFields != null) {
+            builder.startArray(SearchSourceBuilder.FIELDDATA_FIELDS_FIELD.getPreferredName());
+            for (String fieldDataField : fieldDataFields) {
+                builder.value(fieldDataField);
+            }
+            builder.endArray();
+        }
+        if (scriptFields != null) {
+            builder.startObject(SearchSourceBuilder.SCRIPT_FIELDS_FIELD.getPreferredName());
+            for (ScriptField scriptField : scriptFields) {
+                scriptField.toXContent(builder, params);
+            }
+            builder.endObject();
+        }
+        if (sorts != null) {
+            builder.startArray(SearchSourceBuilder.SORT_FIELD.getPreferredName());
+            for (SortBuilder<?> sort : sorts) {
+                sort.toXContent(builder, params);
+            }
+            builder.endArray();
+        }
+        if (highlightBuilder != null) {
+            builder.field(SearchSourceBuilder.HIGHLIGHT_FIELD.getPreferredName(), highlightBuilder, params);
+        }
+        if (query != null) {
+            builder.field(SearchSourceBuilder.QUERY_FIELD.getPreferredName(), query, params);
+        }
+        if (innerHitsBuilder != null) {
+            builder.field(SearchSourceBuilder.INNER_HITS_FIELD.getPreferredName(), innerHitsBuilder, params);
+        }
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        InnerHitBuilder that = (InnerHitBuilder) o;
+        return Objects.equals(name, that.name) &&
+                Objects.equals(nestedPath, that.nestedPath) &&
+                Objects.equals(parentChildType, that.parentChildType) &&
+                Objects.equals(from, that.from) &&
+                Objects.equals(size, that.size) &&
+                Objects.equals(explain, that.explain) &&
+                Objects.equals(version, that.version) &&
+                Objects.equals(trackScores, that.trackScores) &&
+                Objects.equals(fieldNames, that.fieldNames) &&
+                Objects.equals(fieldDataFields, that.fieldDataFields) &&
+                Objects.equals(scriptFields, that.scriptFields) &&
+                Objects.equals(fetchSourceContext, that.fetchSourceContext) &&
+                Objects.equals(sorts, that.sorts) &&
+                Objects.equals(highlightBuilder, that.highlightBuilder) &&
+                Objects.equals(query, that.query) &&
+                Objects.equals(innerHitsBuilder, that.innerHitsBuilder);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(name, nestedPath, parentChildType, from, size, explain, version, trackScores, fieldNames,
+                fieldDataFields, scriptFields, fetchSourceContext, sorts, highlightBuilder, query, innerHitsBuilder);
+    }
+
+    @Override
+    public String toString() {
```

Comment:
Ok, I thought that `ToXContentToBytes` was only used for things that we didn't have a builder for. Makes sense then to extend `ToXContentToBytes`.